torch_ttnn/passes/lowering/add_data_move_pass.py

```diff
@@ -315,252 +288,249 @@ def is_target_a_user_of_curr_node(curr_node, target):
+            if spec.device is None and spec.layout is None and spec.dtype is None:
+                return None
+            return spec
         return None
 
-    g = dst_node.graph
-    new_nodes = []
-    with g.inserting_before(dst_node):
-        if dst_node.target != target_wrappers.repeat:
-            new_nodes.append(
-                g.call_function(ttnn.to_layout, (new_nodes[-1] if new_nodes else src_node, TtnnTileLayout()))
-            )
-        new_nodes.append(
-            g.call_function(ttnn.to_device, (new_nodes[-1] if new_nodes else src_node,), {"device": device})
-        )
-
-    insert_node_between(src_node, dst_idx, dst_node, new_nodes)
-
-    return new_nodes[-1]
-
+    def _change_layout(self, spec, aligning_nodes):
+        g = self.graph
+        need_from_device = False
+        need_to_layout = False
+        need_to_device = False
+        if spec.device == "host":
+            need_from_device = True
+        elif spec.device is not None:
+            need_from_device = True
+            need_to_device = True
+        if spec.layout is not None:
+            need_to_layout = True
 
-def try_add_data_move_out_for_list_args(src_nodes, dst_idx, dst_node):
-    # Handle list type arguments from ops like cat
-    if not isinstance(src_nodes, list):
-        return None
+        if need_from_device:
+            aligning_nodes.append(
+                g.call_function(ttnn.from_device, (aligning_nodes[-1] if aligning_nodes else spec.input_node,))
+            )
+        if need_to_layout:
+            aligning_nodes.append(
+                g.call_function(
+                    ttnn.to_layout, (aligning_nodes[-1] if aligning_nodes else spec.input_node, spec.layout())
+                )
+            )
+        if need_to_device:
+            aligning_nodes.append(
+                g.call_function(
+                    ttnn.to_device,
+                    (aligning_nodes[-1] if aligning_nodes else spec.input_node,),
+                    {"device": spec.device()},
+                )
+            )
 
-    new_nodes = list()
-    for src_node in src_nodes:
-        if should_add_data_move_out(src_node, dst_node):
-            g = dst_node.graph
-            with g.inserting_before(dst_node):
-                target_users_ops = [user.target for user in src_node.users.keys()]
-                to_torch_node = call_to_torch_with_meta(g, src_node)
-                copy_node = try_call_aten__to_copy_with_meta(g, to_torch_node, target_users_ops)
+    def _create_aligned_node(self, spec):
+        aligning_nodes = []
+        g = self.graph
+        if isinstance(spec, self.AlignSpecFromTorch):
+            kwargs = {"layout": TtnnTileLayout(), "device": TtnnDevice()}
+            if spec.dtype is not None:
+                kwargs["dtype"] = spec.dtype()
+            aligning_nodes.append(g.call_function(ttnn.from_torch, (spec.input_node,), kwargs))
+            self._change_layout(spec, aligning_nodes)
```

Comment:
Maybe check if we actually need to change layout (e.g. the target layout is row major or device is host) before calling _change_layout? Otherwise I suspect currently we will insert from_device/to_layout/to_device for all from_torch even if the target layout is tile layout on device

Reply:
Oh yes..., I'll give a check or it will insert from_device/to_layout/to_device for all from_torch just like you said

Reply:
added